pkg/phlaredb/compact.go

```diff
@@ -469,73 +573,36 @@ func (it *dedupeProfileRowIterator) Next() bool {
 	}
 }
 
-func prepareIndexWriter(ctx context.Context, path string, readers []BlockReader) (*index.Writer, error) {
-	var symbols index.StringIter
-	indexw, err := index.NewWriter(ctx, path)
-	if err != nil {
-		return nil, err
-	}
-	for i, r := range readers {
-		if i == 0 {
-			symbols = r.Index().Symbols()
-		}
-		symbols = tsdb.NewMergedStringIter(symbols, r.Index().Symbols())
-	}
-
-	for symbols.Next() {
-		if err := indexw.AddSymbol(symbols.At()); err != nil {
-			return nil, errors.Wrap(err, "add symbol")
-		}
-	}
-	if symbols.Err() != nil {
-		return nil, errors.Wrap(symbols.Err(), "next symbol")
-	}
-
-	return indexw, nil
-}
-
 type symbolsRewriter struct {
-	profiles    iter.Iterator[profileRow]
 	rewriters   map[BlockReader]*symdb.Rewriter
+	w           *symdb.SymDB
 	stacktraces []uint32
-	err         error
 
 	numSamples uint64
 }
 
-func newSymbolsRewriter(it iter.Iterator[profileRow], blocks []BlockReader, w *symdb.SymDB) *symbolsRewriter {
-	sr := symbolsRewriter{
-		profiles:  it,
-		rewriters: make(map[BlockReader]*symdb.Rewriter, len(blocks)),
+func newSymbolsRewriter(path string) *symbolsRewriter {
+	return &symbolsRewriter{
+		w: symdb.NewSymDB(symdb.DefaultConfig().
+			WithDirectory(filepath.Join(path, symdb.DefaultDirName)).
+			WithParquetConfig(symdb.ParquetConfig{
+				MaxBufferRowCount: defaultParquetConfig.MaxBufferRowCount,
+			})),
+		rewriters: make(map[BlockReader]*symdb.Rewriter),
 	}
-	for _, r := range blocks {
-		sr.rewriters[r] = symdb.NewRewriter(w, r.Symbols())
-	}
-	return &sr
 }
 
 func (s *symbolsRewriter) NumSamples() uint64 { return s.numSamples }
 
-func (s *symbolsRewriter) At() profileRow { return s.profiles.At() }
-
-func (s *symbolsRewriter) Close() error { return s.profiles.Close() }
-
-func (s *symbolsRewriter) Err() error {
-	if s.err != nil {
-		return s.err
-	}
-	return s.profiles.Err()
-}
-
-func (s *symbolsRewriter) Next() bool {
-	if !s.profiles.Next() {
-		return false
-	}
+func (s *symbolsRewriter) ReWriteRow(profile profileRow) error {
 	var err error
-	profile := s.profiles.At()
 	profile.row.ForStacktraceIDsValues(func(values []parquet.Value) {
 		s.loadStacktracesID(values)
-		r := s.rewriters[profile.blockReader]
+		r, ok := s.rewriters[profile.blockReader]
+		if !ok {
+			r = symdb.NewRewriter(s.w, profile.blockReader.Symbols())
```

Comment:
I suspect this is where I'm doing something wrong.

Reply:
It looks okay, this is how I'd have done it. I need to take a closer look